sessionctx/variable/session.go

```diff
@@ -60,39 +60,72 @@ var (
 type RetryInfo struct {
 	Retrying               bool
 	DroppedPreparedStmtIDs []uint32
-	currRetryOff           int
-	autoIncrementIDs       []int64
+	autoIncrementIDs       retryInfoAutoIDs
+	autoRandomIDs          retryInfoAutoIDs
 }
 
 // Clean does some clean work.
 func (r *RetryInfo) Clean() {
-	r.currRetryOff = 0
-	if len(r.autoIncrementIDs) > 0 {
-		r.autoIncrementIDs = r.autoIncrementIDs[:0]
-	}
+	r.autoIncrementIDs.clean()
+	r.autoRandomIDs.clean()
+
 	if len(r.DroppedPreparedStmtIDs) > 0 {
 		r.DroppedPreparedStmtIDs = r.DroppedPreparedStmtIDs[:0]
 	}
 }
 
-// AddAutoIncrementID adds id to AutoIncrementIDs.
-func (r *RetryInfo) AddAutoIncrementID(id int64) {
-	r.autoIncrementIDs = append(r.autoIncrementIDs, id)
-}
-
 // ResetOffset resets the current retry offset.
 func (r *RetryInfo) ResetOffset() {
-	r.currRetryOff = 0
+	r.autoIncrementIDs.resetOffset()
+	r.autoRandomIDs.resetOffset()
+}
+
+// AddAutoIncrementID adds id to autoIncrementIDs.
+func (r *RetryInfo) AddAutoIncrementID(id int64) {
+	r.autoIncrementIDs.add(id)
```

Comment:
I think should be directly `r.autoRandomIDs.autoIDs = append(r.autoRandomIDs.autoIDs, id)` here.
AddAutoIncrementID()->add()->append() may be too deep.